keps/sig-api-machinery/1693-warnings/README.md

```diff
@@ -109,104 +112,203 @@ In client-go:
-
+3. color the `Warning:` prefix if the terminal is capable of displaying color
+4. add a `--warnings-as-errors` option to exit with a non-zero exit code after executing the command if warnings were encountered
+ 
 In kube-apiserver and kube-controller-manager, configure the process-wide handler to ignore warnings
 
 ## Design Details
+ 
+### Server-side
 
-Server-side:
+**Warning mechanism:**
 
 * Add a handler chain filter that attaches a WarningRecorder implementation to the request context
-* Add a WarningRecord implementation that deduplicates the warning message per request and writes a `Warning` header
+* Add a WarningRecorder implementation that de-duplicates the warning message per request and writes a `Warning` header
   * the header structure is defined in [RFC2616#14.46](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.46)
-  * warnings would be written with code `299` and warn-agent of `-`
-* In the endpoint installer, decorate handlers for deprecated APIs:
+  * warnings are written with code `299` and warn-agent of `-`
+  * no more than 4K of warning messages are written to avoid exceeding client and proxy header size limits;
+    if the total warning size exceeds that limit, individual warnings are truncated to 256 characters;
+    if the total warning size still exceeds that limit, warnings are output in the order added until the limit is reached;
+    for reference, the longest proposed deprecation warning generated by `kube-apiserver` would be ~170 characters,
+    so ~24 warning messages of that length could be accommodated
+
+**In-process deprecated API warnings:**
+
+In the endpoint installer, decorate handlers for deprecated APIs:
+  * add a warning header: `<group>/<version> <kind> is deprecated in v1.X+, unavailable in v1.Y+; use <replacementGroup>/<replacementVersion> <replacementKind>`
+  * increment the counter metric for the deprecated use
+  * add an audit annotation indicating the request was made to a deprecated API
+
+**In-process validation and admission warnings:**
+
+Enable adding warnings in validation, strategy, and admission handlers:
+* Document when warnings should be surfaced (e.g. when a problematic value is present in a create request,
+  or is introduced by an update request)
+* Define a helper method for adding a field-specific warning that takes a `field.Path` and message
+
+**CustomResourceDefinition version deprecation warnings:**
+
+Add `deprecated bool` and `deprecationWarning string` fields to the `CustomResourceDefinitionVersion` type:
+
+```go
+type CustomResourceDefinitionVersion struct {
+...
+	// deprecated indicates this version of the custom resource API is deprecated.
+	// When set to true, API requests to this version receive a warning header in the server response,
+	// and an audit annotation indicating a deprecated API was used is added to the audit event for the request.
+	// Defaults to false.
+	// +optional
+	Deprecated bool `json:"deprecated,omitempty"`
```

Comment:
I think we can argue about this in the PR, but I think it's useful to allow well formed versions independent of kube versions

Reply:
DeprecationWarning message allows surfacing any version/replacement text desired